src/cobra/core/gene.py

```diff
@@ -559,8 +565,156 @@ def _repr_html__(self):
+
+        Returns
+        ------
+        Symbol or BooleanFunction
+            SYMPY expression (Symbol or And or Or). Symbol("") if the GPR is empty
+
+        Notes
+        -----
+        Calls _symbolic_gpr()
+        """
+        # noinspection PyTypeChecker
+        if names:
+            GPRGene_dict = {gid: Symbol(names[gid]) for gid in self.genes}
+        else:
+            GPRGene_dict = None
+        return self._symbolic_gpr(self, GPRGene_dict=GPRGene_dict)
+
+    def _symbolic_gpr(
+        self,
+        expr: Union["GPR", Expression, BoolOp, Name, list] = None,
+        GPRGene_dict: dict = None,
+    ) -> Union[spl.Or, spl.And, Symbol]:
+        """Parse gpr into SYMPY using ast similar to _ast2str().
+
+        Parameters
+        ----------
+        expr : AST or GPR or list or Name or BoolOp
+            compiled GPR
+        GPRGene_dict: dict
+            dictionary from gene id to GPRGeneSymbol
+        Returns
+        -------
+        Symbol or BooleanFunction
+            SYMPY expression (Symbol or And or Or). Symbol("") if the GPR is empty
+        """
+        if GPRGene_dict is None:
+            GPRGene_dict = {gid: Symbol(name=gid) for gid in expr.genes}
+        if isinstance(expr, (Expression, GPR)):
+            return (
+                self._symbolic_gpr(expr.body, GPRGene_dict) if expr.body else Symbol("")
+            )
+        else:
+            if isinstance(expr, Name):
+                return GPRGene_dict.get(expr.id)
+            elif isinstance(expr, BoolOp):
+                op = expr.op
+                if isinstance(op, Or):
+                    # noinspection PyTypeChecker
+                    sym_exp = spl.Or(
+                        *[self._symbolic_gpr(i, GPRGene_dict) for i in expr.values]
+                    )
+                elif isinstance(op, And):
+                    # noinspection PyTypeChecker
+                    sym_exp = spl.And(
+                        *[self._symbolic_gpr(i, GPRGene_dict) for i in expr.values]
+                    )
+                else:
+                    raise TypeError("Unsupported operation " + op.__class__.__name)
+                return sym_exp
+            elif not expr:
+                return Symbol("")
+            else:
+                raise TypeError("Unsupported Expression  " + repr(expr))
+
+    @classmethod
+    def from_symbolic(cls, sympy_gpr: Union[spl.BooleanFunction, Symbol]) -> "GPR":
+        """Construct a GPR from a sympy expression.
+
+        Parameters
+        ----------
+        sympy_gpr: sympy
+            a sympy that describes the gene rules, being a Symbol for single genes
+            or a BooleanFunction for AND/OR relationships
+
+        Returns
+        -------
+        GPR:
+            returns a new GPR while setting  self.body as
+            Parsed AST tree that has the gene rules
+            This function also sets self._genes with the gene ids in the AST
+
+        """
+
+        def _sympy_to_ast(
```

Comment:
Very cool that you actually went through the hassle of parsing the tree. That should be pretty fast.

Reply:
Thank you. The only other alternative would have been to print SYMPY as str and then parse the str as ast which would have been slow.